components/vocabularies/api/src/main/java/org/phenotips/vocabulary/internal/solr/DefaultVocabularySourceRelocationService.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.vocabulary.internal.solr;
+
+import org.phenotips.Constants;
+import org.phenotips.vocabulary.VocabularySourceRelocationService;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.EntityReference;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * Default implementation for the {@link VocabularySourceRelocationService} component.
+ *
+ * @version $Id$
+ * @since 1.4
+ */
+@Component
+@Singleton
+public class DefaultVocabularySourceRelocationService implements VocabularySourceRelocationService
+{
+    /** The document where vocabulary relocation sources are stored. */
+    private static final EntityReference VOCABULARY_RELOCATION_CLASS = new EntityReference(
+        "VocabularySourceRelocationClass", EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+
+    /** The document where preferences are stored. */
+    private static final EntityReference PREFERENCES_DOCUMENT =
+        new EntityReference("XWikiPreferences", EntityType.DOCUMENT, Constants.XWIKI_SPACE_REFERENCE);
+
+    @Inject
+    private Provider<XWikiContext> provider;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public String getRelocation(String original)
```

Comment:
@sdumitriul I discussed this with @veronikaslc and she said you should know better: why so we pass URLs around in the new design? I'e. when a vocabulary needs to be reindexed, don't we want to reindex "vocabulary named X" and not "vocabulary which used to be indexed from URL http://x.y.z"? It seems in this design we always need to remember the old URL, instead of just new URL for each vocabulary type?